pkg/graveler/committed/merge_iterator.go

```diff
@@ -68,75 +81,135 @@ func (d *compareIterator) handleConflict() bool {
 		d.err = graveler.ErrConflictFound
 		return false
 	}
-	d.val = d.diffIt.Value().Copy()
+	val, _ := d.diffIt.Value()
+	d.val = val.Copy()
 	d.val.Type = graveler.DiffTypeConflict
 	return true
 }
 
-func (d *compareIterator) Next() bool {
-	for {
-		// get next value or return with context/iterator error (if any)
+func (d *compareIterator) Step() bool {
+	hasNext := true
+	for hasNext {
 		select {
 		case <-d.ctx.Done():
 			d.err = d.ctx.Err()
 			return false
 		default:
-			if !d.diffIt.Next() {
-				d.err = d.diffIt.Err()
-				return false
-			}
-		}
-		val := d.diffIt.Value()
-		key := val.Key
-		typ := val.Type
-		baseVal, err := d.valueFromBase(key)
-		if err != nil {
-			d.err = err
-			return false
 		}
-		switch typ {
-		case graveler.DiffTypeAdded:
-			// exists on source, but not on dest
-			if baseVal == nil {
-				// added only on source
-				d.val = d.diffIt.Value().Copy()
-				return true
-			}
-			if !bytes.Equal(baseVal.Identity, val.Value.Identity) {
-				// removed on dest, but changed on source
-				return d.handleConflict()
-			}
-			continue
-		case graveler.DiffTypeChanged:
-			if baseVal == nil {
-				// added on dest and source, with different identities
-				return d.handleConflict()
+		val, rngDiff := d.diffIt.Value()
+		if rngDiff != nil {
+			typ := rngDiff.Type
+			rng := rngDiff.Range
+			baseRange, err := d.rangeFromBase(graveler.Key(rng.MinKey))
+			if err != nil {
+				d.err = err
+				return false
 			}
-			if bytes.Equal(baseVal.Identity, val.Value.Identity) {
-				// changed on dest, but not on source
-				continue
+			switch typ {
+			case graveler.DiffTypeAdded:
+				// exists on source, but not on dest
```

Comment:
Can you document `DiffTypeAdded` better?  I would expect this to be "exists on dest, but not on source" -- here it's the reverse?!  (Maybe add pretty pictures...?)

Reply:
This also wasn't added as part of this PR, the reason is that we are holding a diff iterator that does a 2 way diff between destination and source. I could also change it if it makes more sense. prefer not as part of this PR. @arielshaqed what do you think?